internal/mode/static/nginx/config/http/config.go

```diff
@@ -86,3 +86,8 @@ type MapParameter struct {
 	Value  string
 	Result string
 }
+
+const (
+	// InternalLocationPrefix indicates that the path is internal type.
+	InternalLocationPrefix string = "/ngf-internal"
```

Comment:
per offline discussion, let's change this prefix to `ngf-internal` (remove the `/`). This way, we don't need to validate and require users not start paths in HTTPRoutes with that prefix, as it will impossible to send a request to NGINX with a URI without `/` at the beginning.

So once we update, we can remove the validation and the corresponding docs.

Reference configuration example that proves the approach:

```nginx
split_clients $request_id $dest {
    50.0% dest-1;
    50.0% dest-2;
}

server {
    listen 8080;

    location / {
        rewrite ^ $dest last;
    }

    location dest-1 {
        internal;
        rewrite ^ $request_uri;
        return 200 "dest-1:$uri\n";
    }

    location dest-2 {
        internal;
        rewrite ^ $request_uri;
        return 200 "dest-2:$uri\n";
    }

    # doesn't collide with dest-1
    location /dest-1 {
        return 200 "/dest-1\n";
    }
    # doesn't collide with dest-2
    location /dest-2 {
        return 200 "/dest-2\n";
    }
}
```

 NGINX rejects requests with URI that don’t start with / so this means location like ‘dest-1’ can’t be accessed from outside.
```
(using telnet)
GET test HTTP/1.1
HTTP/1.1 400 Bad Request
```
Logs:
```
2024/01/10 15:51:42 [info] 1168#1168: *22 client sent invalid request while reading client request line, client: 127.0.0.1, server: , request: "GET test HTTP/1.1"
```

Reply:
During further offline discussions, it was suggested to use named locations.

So instead of 
```
location /ngf-internal-rule5-route0 {
  internal;
  . . .
  }
```

we can use:
```
location @rule5-route0 {
   ...
}
```

again, this way, no restrictions for path values in HTTPRoute rules are needed

Reply:
@salonichf5  

this wasn't implemented -- https://github.com/nginxinc/nginx-gateway-fabric/pull/1445#discussion_r1447949407 

but I'm kind of wonder now if we do it in a separate PR, cause that will change the approach slightly - internal location -> named location.

@sjberman @kate-osborn  thoughts?

Reply:
Ultimately isn't that change the one that we decide fixes the bug? If we don't do that, what would this PR really be for?

Reply:
> Ultimately isn't that change the one that we decide fixes the bug? If we don't do that, what would this PR really be for?

the current code fixes the bug too.
however, the named location approach is better, because named location are internal-only locations.

so with the current aproach, we have an internal location:
```
location rule5-route0 {
  internal;
  . . .
  }
```
and we use a "hack" by not using `/` in the path. This prevents the location from colliding with any paths in HTTPRoutes (because paths in HTTPRoutes must start with '/'). More over, NGINX will reject any requests with 400 error code with paths _without_ '/'. 

With named locations, we use

location @rule5-route0 {
   ...
}

such locations are not acessible from the outside by design and they cannot collide with any paths in HTTPRoutes because of that.
so this solution is less hacky.
However, it will also require a few extra changes. For example, we will no longer need to do this https://github.com/nginxinc/nginx-gateway-fabric/blob/9d9c1f249b2c55d23e86c226bb289965aeff8e47/internal/mode/static/nginx/config/servers.go#L322 , because jumping to an internal location doesn't change the request_uri

Reply:
Personally, I would just make those changes in this PR, but I'm also OK with a separate PR. Whatever is easier.

Reply:
@pleshakov Ok, we'll have to do some testing with the rewrites after these changes then to ensure that they still work. Specifically when using HTTP match conditions with a rewrite, to trigger this specific case.

Reply:
kk.
let's proceed with that change then.

this requires the following:
- adding `@` to the generated paths for internal location. So that instead of `rule5-route0` it will be `@rule5-route0`. This will make that location a named location (for reference, named locations are described here https://nginx.org/en/docs/http/ngx_http_core_module.html#location )
- because named locations don't require InternalRewrite, we need to remove this field https://github.com/nginxinc/nginx-gateway-fabric/blob/9d9c1f249b2c55d23e86c226bb289965aeff8e47/internal/mode/static/nginx/config/servers.go#L97 and rename MainRewrite to just Rewrite. The corresponding code that uses InternalRewrite field will need to be removed too.

Reply:
I'll make the necessary changes, I misunderstood the expectation

Reply:
I've made the updates to code but i don't think the necessary location blocks are being created, would love some feedback @sjberman @pleshakov @kate-osborn

Reply:
@salonichf5 

the conformance tests that are failing are related to query params matching

let's try to remove this part https://github.com/nginxinc/nginx-gateway-fabric/blame/9d9c1f249b2c55d23e86c226bb289965aeff8e47/internal/mode/static/nginx/modules/src/httpmatches.js#L49-L54 that was added in https://github.com/nginxinc/nginx-gateway-fabric/commit/0f465a75f007c8853bf08ddbe2b8241006381be4#diff-a3f234c23f4767dddb8aee393ba5696057eedfe3565f94effced0436d46b8e13 , as it changes internal location to something from `@rule5-route0` to `@rule5-route0?a=b`, which probably breaks internal redirect logic.

cc @sjberman

Reply:
updated

Reply:
Pipeline passes, but I still don't see the correct location blocks being created. 

are internal locations not supposed to be visible in the config?

I don't see any locations being created will try to debug.

Reply:
I was able to run verify that these two examples 

1. https://docs.nginx.com/nginx-gateway-fabric/how-to/traffic-management/advanced-routing/
2. https://github.com/nginxinc/nginx-gateway-fabric/blob/main/site/content/how-to/traffic-management/redirects-and-rewrites.md 

Thanks to @kate-osborn 

Please let me know if I need to verify using other examples as well. @sjberman @pleshakov